Withings authentication and webhook fixes

vangorra opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Fix bug that resulted in constantly re-authenticating withings.
Fix bug that broke withings webhooks when hass domain changed or users enabled/disabled nabu.casa.
Fix bug that resulted in multiple reauth flows being created for one withings user.
Clear withings reauth flow for user if subsequent api calls are successful.
Workaround issue where withings will stop calling the hass webhook by polling always for data every 30 minutes.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #37290, #42717, #41601
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-asc+-review%3Aapproved

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

frenck commented:
I'm confused, you need to save the webhook in the config entries as you'd need to re-use it later on?

How is removing it prevent issues with the cloud hooks?

vangorra commented:
Previously, we stored both the webhook_id and generated webhook_url in the config. This resulted in webhooks failing when users changed the external URL or enabled/disabled nabu.casa. Even though the component's webhook sync with Withings occurs every 2 hours (and on startup), the user's changes would have no effect because the config's webhook_url is what was sent to Withings.

This change only persists the webhook_id. When the component's webhook sync runs, it will generate the webhook url and pass that to Withings. This ensures Withings is regularly updated with the correct webhook url instead of a persisted one.

frenck commented:
It should not generate a new one... that is the whole point.

See for example Netatmo or Toon integrations on how to handle this correctly.

vangorra commented:
@frenck I see both those integrations persist the cloudhook url in the config entry. Why is that? What if the user deletes their cloud account and creates a new one? How would we update the URL for the new cloud account? I tend to think that generating a new webhook/cloudhook on config entry setup would be most flexable.

frenck commented:
> I tend to think that generating a new webhook/cloudhook on config entry setup would be most flexable.

That is simply not allowed, it will a lot of webhooks, that is not needed.

> I see both those integrations persist the cloudhook url in the config entry.

Yes, it should.

> Why is that? What if the user deletes their cloud account and creates a new one? How would we update the URL for the new cloud account? 

It stores the webhook ID, not the URL. Hence it is not an issue. The ID is used to generate the URL, which means the registered webhook is re-used. For cloud it will result in the same URL (unless account is changed). For local, it will result in the same URL (unless URL settings are changed).

vangorra commented:
It makes sense the webhook id would be persisted but I'm confused why we would persist the cloudhook url. The code below shows the cloudhook url is persisted. This example is consistent with toon and netatmo. Shouldn't the webhook id be persisted by the cloudhook url and/or webhook url be regenerated on startup?

https://github.com/home-assistant/core/blob/dev/homeassistant/components/toon/coordinator.py#L55-L72
```python
        if CONF_WEBHOOK_ID not in self.entry.data:
            data = {**self.entry.data, CONF_WEBHOOK_ID: secrets.token_hex()}
            self.hass.config_entries.async_update_entry(self.entry, data=data)

        if self.hass.components.cloud.async_active_subscription():

            if CONF_CLOUDHOOK_URL not in self.entry.data:
                webhook_url = await self.hass.components.cloud.async_create_cloudhook(
                    self.entry.data[CONF_WEBHOOK_ID]
                )
                data = {**self.entry.data, CONF_CLOUDHOOK_URL: webhook_url}
                self.hass.config_entries.async_update_entry(self.entry, data=data)
            else:
                webhook_url = self.entry.data[CONF_CLOUDHOOK_URL]
        else:
            webhook_url = self.hass.components.webhook.async_generate_url(
                self.entry.data[CONF_WEBHOOK_ID]
            )
```

vangorra commented:
@frenck Could you provide some clarity on the comment above? The examples you mention both persist the cloudhook url. If you change nabu casa accounts, these integrations would need to be removed and added in order for them to use a different account.

vangorra commented:
@frenck Could you provide some clarification to the above comment?

vangorra committed:
Fix bug where withings neded constant re-authentication.
Workaround issue where withings webhooks don't notify hass.
Fix bug that created multiple withings reauth flows.
Clear withings reauth flow if api returns successfully. Withings occasionally (erronousally) returns invalid creds.
Added options flow to enable/disable withings webhook support.

vangorra committed:
Poll for withings data every 30 minutes despise webhook config.

vangorra commented:
Ready for review.

vangorra committed:
Removing invalidation_version from yaml config.

bdraco commented on homeassistant/components/withings/common.py:
```diff
@@ -1018,69 +1097,165 @@ async def async_added_to_hass(self) -> None:
+        is_cloud = hass.components.cloud.async_active_subscription()
+        if is_cloud:
+            if not cloudhook_url:
+                # Delete the existing hook (if exists) because don't have the URL.
+                # try:
+                #     await hass.components.cloud.async_delete_cloudhook(webhook_id)
+                # except ValueError:
+                #     pass
+
```
Commented code

frenck commented:
> @frenck Could you provide some clarity on the comment above? The examples you mention both persist the cloudhook url. If you change nabu casa accounts, these integrations would need to be removed and added in order for them to use a different account.

Yes, at this point, you are correct. However, that is a small case, which should not be solved on the integration level. For the future, we should look into getting a helper that does all this magic (for both cloud and local).

pdecat commented:
I believe this is superseded by https://github.com/home-assistant/core/pull/47975
